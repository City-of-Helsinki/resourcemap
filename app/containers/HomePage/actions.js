import {
  LOAD_RESOURCE,
  LOAD_RESOURCE_COMPLETED,
  LOAD_RESOURCE_ERROR,
} from './constants';

export function loadResource() {
  return {
    type: LOAD_RESOURCE,
  };
}

export function loadResourceCompleted(resource) {
  return {
    type: LOAD_RESOURCE_COMPLETED,
    resource,
  };
}

export function loadResourceError() {
  return {
    type: LOAD_RESOURCE_ERROR,
  };
}
