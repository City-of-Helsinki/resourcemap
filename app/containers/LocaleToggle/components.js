import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const Wrapper = styled.div`
  padding: 2px;
`;

export const LanguageList = styled.ul`
  margin: 0;
  padding: 0;
  display: flex;

  list-style: none;
`;

const LanguageItem = styled.li`
  padding: 0;

  font-size: ${props => props.theme.fontSize.get(2)};
  font-weight: bold;

  &:not(:last-child) {
    margin-right: 20px;
  }
`;

const LanguageButton = styled.button`
  padding: 0;
  margin: 0;

  color: inherit;
  font-weight: inherit;

  background: none;
  border: none;
`;

export const LanguageOption = ({ onClick, children }) => (
  <LanguageItem>
    <LanguageButton type="button" onClick={onClick}>
      {children}
    </LanguageButton>
  </LanguageItem>
);

LanguageOption.propTypes = {
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
};
