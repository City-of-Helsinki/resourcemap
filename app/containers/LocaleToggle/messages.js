/*
 * LocaleToggle Messages
 *
 * This contains all the text for the LanguageToggle component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
	en: {
		id: 'boilerplate.containers.LocaleToggle.en',
		defaultMessage: 'In English',
	},
	de: {
		id: 'boilerplate.containers.LocaleToggle.de',
		defaultMessage: 'de',
	},
	fi: {
		id: 'boilerplate.containers.LocaleToggle.de',
		defaultMessage: 'Suomeksi',
	},
	sv: {
		id: 'boilerplate.containers.LocaleToggle.de',
		defaultMessage: 'PÃ¥ svenska',
	},
});
