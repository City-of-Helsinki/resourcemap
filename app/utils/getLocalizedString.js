import get from 'lodash/get';

import DEFAULT_LOCALE from '../i18n';

function getLocalizedString(localizationObject, locale = DEFAULT_LOCALE) {
  if (!localizationObject) {
    return '';
  }

  const localizedString = get(localizationObject, locale, null);

  // If we do not find a localized string, we'll try to use the default
  // locale.
  if (localizedString === null) {
    return get(localizationObject, DEFAULT_LOCALE);
  }

  return localizedString;
}
export default getLocalizedString;
