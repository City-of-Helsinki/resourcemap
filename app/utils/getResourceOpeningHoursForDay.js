import dateFormat from 'dateformat';

function getOpeningHoursForDay(openingHours, day) {
  return openingHours.find(openingHour => {
    const date = dateFormat(day, 'yyyy-mm-dd');
    const openingHourDate = dateFormat(
      new Date(openingHour.date),
      'yyyy-mm-dd',
    );

    return date === openingHourDate;
  });
}

export default getOpeningHoursForDay;
