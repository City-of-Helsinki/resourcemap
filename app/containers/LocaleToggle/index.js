/*
 *
 * LanguageToggle
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import { FormattedMessage } from 'react-intl';

import { Wrapper, LanguageList, LanguageOption } from './components';
import messages from './messages';
import { appLocales } from '../../i18n';
import { changeLocale } from '../LanguageProvider/actions';
import { makeSelectLocale } from '../LanguageProvider/selectors';

export function LocaleToggle({ currentLocale, onLocaleToggle }) {
  return (
    <Wrapper>
      <LanguageList>
        {appLocales
          .filter(locale => locale !== currentLocale)
          .map(locale => (
            <LanguageOption key={locale} onClick={() => onLocaleToggle(locale)}>
              <FormattedMessage {...messages[locale]} />
            </LanguageOption>
          ))}
      </LanguageList>
    </Wrapper>
  );
}

LocaleToggle.propTypes = {
  onLocaleToggle: PropTypes.func,
  currentLocale: PropTypes.string,
};

const mapStateToProps = createSelector(
  makeSelectLocale(),
  locale => ({
    currentLocale: locale,
  }),
);

export function mapDispatchToProps(dispatch) {
  return {
    onLocaleToggle: locale => dispatch(changeLocale(locale)),
    dispatch,
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(LocaleToggle);
