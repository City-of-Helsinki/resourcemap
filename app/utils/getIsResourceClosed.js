import get from 'lodash/get';
import defaultTo from 'lodash/defaultTo';

import getResourceOpeningHoursForDay from 'utils/getResourceOpeningHoursForDay';

function getOpensAndCloses(resource, date) {
  const openingHours = get(resource, 'opening_hours', []);
  const openingHoursForDate = getResourceOpeningHoursForDay(openingHours, date);

  const opens = get(openingHoursForDate, 'opens');
  const closes = get(openingHoursForDate, 'closes');

  return [opens, closes];
}

export function getIsResourceClosedForDay(resource, date) {
  const [opens, closes] = getOpensAndCloses(resource, date);

  return opens === null && closes === null;
}

function getIsResourceClosedNow(opens, closes, date) {
  const opensTime = new Date(opens).getTime();
  const closesTime = new Date(closes).getTime();
  const nowTime = date.getTime();

  const isBeforeOpening = nowTime < opensTime;
  const isAfterClosing = nowTime > closesTime;

  return isBeforeOpening || isAfterClosing;
}

function getResourceIsClosed(resource, dateOverride) {
  const date = defaultTo(dateOverride, new Date());

  if (getIsResourceClosedForDay(resource, date)) {
    return true;
  }

  const [opens, closes] = getOpensAndCloses(resource, date);

  if (opens && closes) {
    return getIsResourceClosedNow(opens, closes, date);
  }

  return false;
}

export default getResourceIsClosed;
