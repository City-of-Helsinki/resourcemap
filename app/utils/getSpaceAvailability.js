import isResourceAvailable from 'utils/isResourceAvailable';
import getIsResourceClosed from 'utils/getIsResourceClosed';
import SpaceAvailability from 'constants/SpaceAvailability';

function getSpaceAvailability(space) {
  if (!space.data) {
    return SpaceAvailability.NO_DATA;
  }

  if (!space.canBeReserved || getIsResourceClosed(space.data)) {
    return SpaceAvailability.CLOSED;
  }

  const isAvailable = isResourceAvailable(new Date(), space.data);

  if (isAvailable) {
    return SpaceAvailability.AVAILABLE;
  }

  return SpaceAvailability.TAKEN;
}

export default getSpaceAvailability;
