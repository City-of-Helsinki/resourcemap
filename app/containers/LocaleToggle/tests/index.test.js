import React from 'react';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';
import { render } from 'tests/utils';

import LocaleToggle, { mapDispatchToProps } from '../index';
import { changeLocale } from '../../LanguageProvider/actions';
import LanguageProvider from '../../LanguageProvider';

import configureStore from '../../../configureStore';
import { translationMessages } from '../../../i18n';

describe('<LocaleToggle />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should match the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <LanguageProvider messages={translationMessages}>
          <LocaleToggle />
        </LanguageProvider>
      </Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not present the default `fi` Finnish language option', () => {
    const { queryByText } = render(
      <Provider store={store}>
        <LanguageProvider messages={translationMessages}>
          <LocaleToggle />
        </LanguageProvider>
      </Provider>,
    );

    expect(queryByText('Suomeksi')).toBeNull();
  });

  describe('mapDispatchToProps', () => {
    describe('onLocaleToggle', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onLocaleToggle).toBeDefined();
      });

      it('should dispatch changeLocale when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const locale = 'sv';
        result.onLocaleToggle(locale);
        expect(dispatch).toHaveBeenCalledWith(changeLocale(locale));
      });
    });
  });
});
