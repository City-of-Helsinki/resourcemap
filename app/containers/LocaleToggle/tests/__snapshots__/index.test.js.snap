// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocaleToggle /> should match the snapshot 1`] = `
<div
  class="components__Wrapper-sc-8efs89-0 iFEOpV"
>
  <ul
    class="components__LanguageList-sc-8efs89-1 kOkdRt"
  >
    <li
      class="components__LanguageItem-sc-8efs89-2 iZcJuQ"
    >
      <button
        class="components__LanguageButton-sc-8efs89-3 jjJqum"
        type="button"
      >
        <span>
          In English
        </span>
      </button>
    </li>
    <li
      class="components__LanguageItem-sc-8efs89-2 iZcJuQ"
    >
      <button
        class="components__LanguageButton-sc-8efs89-3 jjJqum"
        type="button"
      >
        <span>
          PÃ¥ svenska
        </span>
      </button>
    </li>
  </ul>
</div>
`;
